text/XXX-unwind-FFI.md

```diff
@@ -0,0 +1,286 @@
+```
+
+`panic.runtime` may only be specified if the panic strategy is `unwind`.
+And just as there may only be one strategy
+in the dependency graph for a final product,
+there may only be one `runtime`.
+
+For now, the only valid `runtime` keys will be:
+
+* `default` - identical to `panic = 'unwind'` with no `runtime` selected.
+  This will use Rust's default unwinding runtime,
+  unless another crate in the final product's dependency graph
+specifies `panic = 'abort'`,
+  in which case the `abort` strategy will take precedence as usual.
+  which is not guaranteed to be compatible with native exceptions.
+* `native` - as discussed above, this will preserve the behavior of the current
+  implementation of `libpanic_unwind`.
+* `crate` - indicates that a non-`std` crate will use `#![panic_runtime]` to
+  provide the runtime.
+* `self` - indicates that this crate itself provides `#![panic_runtime]`.
+
+If the `native` or `crate` key is specified
+anywhere in a final product's dependency graph,
+no crate in that dependency graph may specify the `panic = abort` strategy;
+this mismatch will cause the build to fail.
+
+The function annotation `#[unwind(allowed)]` will only be permitted in crates
+that specify a non-default `panic.runtime`.
+
+Crates that do not use the `profile.foo.panic` option at all
+will remain compatible with any `profile.foo.panic` configuration
+used to generate the final product.
+
+For non-Cargo users, equivalent `rustc` flags will be provided
+(which will be how Cargo itself implements the option).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Unwinding for functions with the `#[unwind(allowed)]` annotation
+is performed as if the function were not marked `extern`.
+This annotation is not permitted on functions not marked `extern`.
+It has no observable effect unless the marked function `panic`s
+(e.g. it has no observable effect
+when a function returns normally or enters an infinite loop).
+The LLVM IR for such functions must not be marked `nounwind`.
+
+The compiler will have a new stable flag, `-C panic.runtime`,
+which will be required to enable the `#[unwind(allowed)]` annotation;
+as explained above, the flag will specify the expected source
+of the panic runtime to be used in the final (binary) product.
+If the source is `default`,
+the `libpanic_unwind` crate will provide the runtime,
+and the `#[unwind(allowed)]` annotation will not be permitted in that crate.
+
+Different values of `panic.runtime` will not be permitted for different crates
+in a single dependency graph,
+with the exception of `self`, which may only be used once
+in a dependency graph,
+and is only compatible with crates using the `crate` value.
+Because the `panic` runtime is lazily selected and linked,
+crates that do not specify a value for `panic.strategy`
+are compatible with crates using any of the four values.
+Crates that explicitly specify the `default` runtime, however,
```

Comment:
This essentially makes such crates incompatible with `libstd`/`libcore` etc, which come precompiled and use the `crate` panic strategy AFAIR. Using such crates realistically would require user to compile their own `libstd` or `libcore` – or some other mechanism for distribution will need to happen (such as metadata-only-rlibs).

Reply:
RFC 1513 says:

> The standard library will need a runtime and will be lazily linked to a crate which has `#![panic_runtime]`.

I thought the "lazily" was intended to imply that the standard library could be used with any panic runtime; am I misunderstanding the RFC? I intended the wording of my RFC to use the same "laziness" strategy to ensure compatibility between all `unwind` crates that don't specify _conflicting_ strategies.